nemo/collections/nlp/models/information_retrieval/bert_joint_ir_model.py

```diff
@@ -0,0 +1,240 @@
+        Lightning calls this inside the validation loop with the data from the validation dataloader
+        passed in as `batch`.
+        """
+        input_ids, input_mask, input_type_ids, query_id, passage_ids = batch
+        batch_size, num_passages, seq_length = input_ids.size()
+
+        scores = self(
+            input_ids=input_ids.view(-1, seq_length),
+            token_type_ids=input_type_ids.view(-1, seq_length),
+            attention_mask=input_mask.view(-1, seq_length),
+        ).view(batch_size, 1, num_passages)
+
+        data_for_val = {
+            "scores": scores[0, 0, :].detach().cpu().numpy(),
+            "query_id": query_id.item(),
+            "passage_ids": passage_ids[0].detach().cpu().numpy(),
+        }
+        return data_for_val
+
+    def validation_epoch_end(self, outputs):
+        """
+        Called at the end of validation to aggregate outputs.
+        :param outputs: list of individual outputs of each validation step.
+        """
+
+        query2passages = {}
+        for val_output in outputs:
+            query_id = val_output["query_id"]
+
+            if query_id in query2passages:
+                query2passages[query_id]["psg_ids"] = np.concatenate(
+                    (query2passages[query_id]["psg_ids"], val_output["passage_ids"])
+                )
+                query2passages[query_id]["scores"] = np.concatenate(
+                    (query2passages[query_id]["scores"], val_output["scores"])
+                )
+            else:
+                query2passages[query_id] = {"psg_ids": val_output["passage_ids"], "scores": val_output["scores"]}
+
+        rrs = self.calculate_reciprocal_ranks(query2passages, self._val_qrels)
+
+        tensorboard_logs = {
+            "model_mrr": np.mean(rrs["model"]),
+            "oracle_mrr": np.mean(rrs["oracle"]),
+        }
+
+        return {"log": tensorboard_logs}
+
+    def setup_training_data(self, train_data_config: Optional[DictConfig]):
+        self._train_dl = self._setup_dataloader_from_config(cfg=train_data_config, mode="train")
+
+    def setup_validation_data(self, val_data_config: Optional[DictConfig]):
+        self._validation_dl = self._setup_dataloader_from_config(cfg=val_data_config, mode="eval")
+        self._val_qrels = self.parse_qrels(val_data_config.qrels)
+
+    def setup_test_data(self, test_data_config: Optional[DictConfig]):
+        self._test_dl = self._setup_dataloader_from_config(cfg=test_data_config, mode="eval")
+
+    @staticmethod
+    def parse_qrels(qrels):
+        query2rel = {}
+        for line in open(qrels, "r").readlines():
+            query_id = int(line.split("\t")[0])
+            psg_id = int(line.split("\t")[2])
+            if query_id not in query2rel:
+                query2rel[query_id] = [psg_id]
+            else:
+                query2rel[query_id].append(psg_id)
+        return query2rel
+
+    @staticmethod
+    def calculate_reciprocal_ranks(query2passages, query2rel):
```

Comment:
due to similarity between bert_joint_ir_model and bert_dpr_model
could you make a parent class with common functions?